### ISSUE TITLE
Detecting a variable when it changes failed

### ISSUE THREAD

**Issue by yiakwy** [mobxjs/mobx-react#383]
# Problem 1
Hi @mweststrate , here is the first example for demonstration 
```js
class ObservableTodoStore {
    @observable todos = [];
    @observable pendingRequests = 0;
    @observable status = {changed:false};

    constructor() {
        mobx.autorun(() => console.log(this.report));
        mobx.autorun(() => console.log(this.isChanged));
    }

	@computed get completedTodosCount() {
        // status.changed = true;
      	console.log("calling completedTodosCount");
      	status.changed = !status.changed;
    	return this.todos.filter(
			todo => todo.completed === true
		).length;
    }

	@computed get isChanged(){
    	let ret = status.changed === true ? "is":"is not"
        console.log(`status ${ret} changed`)
        return status.changed
    }

	@computed get report() {
		if (this.todos.length === 0)
			return "<none>";
		return `Next todo: "${this.todos[0].task}". ` +
			`Progress: ${this.completedTodosCount}/${this.todos.length}`;
	}

	addTodo(task) {
		this.todos.push({
			task: task,
			completed: false,
			assignee: null
		});
	}
}


const observableTodoStore = new ObservableTodoStore();
```
## expected result
I want to tell when a variable in tracked function is changed (not called) and trigger a function accordingly. In the above example, I use a boolean variable to pass info into an automatically calling function when it needs to be executed.

## actual result
However here is output:
```bash
[Log] <none> (script.js, line 80)
[Log] status is not changed (script.js, line 80)
[Log] undefined (script.js, line 80)
[Log] calling completedTodosCount (script.js, line 80)
[Log] Next todo: "read MobX tutorial". Progress: 0/1 (script.js, line 80)
[Log] calling completedTodosCount (script.js, line 80)
[Log] Next todo: "read MobX tutorial". Progress: 0/2 (script.js, line 80)
[Log] calling completedTodosCount (script.js, line 80)
[Log] Next todo: "read MobX tutorial". Progress: 1/2 (script.js, line 80)
[Log] Next todo: "grok MobX tutorial". Progress: 1/2 (script.js, line 80)
```
This above function "isChanged" is called only once.

# Problem 2
See I have a Three.js function providing interface to modify a variable "mode"
In react I have React Component to modify the variable value provided by a store, 
like this:

```jsx
@inject("store") @observer
export default class Demo extends React.Component {
	constructor() {
		super()
		this.onModeChoiceChange = this.onModeChoiceChange.bind(this)
	}
...
	render() {
		const { options } = this.props.store

		return (
			<div className="panel canvasPlugin col-2">
				<h3 className="Title"> Control Panle </h3>
				<hr />
				<ul className="sidebar">
				</ul>
				<li className="btn">
				<input className="inp"
						type="checkbox"
						checked={ options.modeChoice.isFirstPersonControls }
			            onClick={ this.onModeChoiceChange } />
				Change Mode from FirstPersonControls 
				</li>
			</div>
		)
	}

	onModeChoiceChange() {
		const { options } = this.props.store
		options.modeChoice.isFirstPersonControls = !options.modeChoice.isFirstPersonControls
	}

// store file in some package
BLA BLA BLA
// delegate attributes of target updated automatically, reactively by mobx
class ObservableSceneStore {
    constructor() {
        autorun(() => console.log(this.ModeChoice));
        autorun(() => console.log(this.updateChoice()));
    }

	@observable styles = {
		width: window.innerWidth,
		height: window.innerHeight,
	}
	
	@observable options = {
		modeChoice : {
			isFirstPersonControls: false,
			codeMap : {
				"TrackballControls" : 0,
				"FirstPersonControls" : 1
			}
		}
	}

	@computed get ModeChoice() {
                console.log("test ModeChoice")
		const mode = options.modeChoice.isFirstPersonControls
		prompt(`Mode choice is ${ mode  === true ? "is" : "is not"} firstPersonControls`)
BLA BLA BLA
                calling some GL programme interface
		return mode
	}
       
        updateChoice() {
                console.log("test updateChoice")
                let options  = this.options,
                     mode = options.modeChoice.isFirstPersonControls ? options.modeChoice.codeMap["FirstPersonControls"]  : options.modeChoice.codeMap["TrackballControls"]
BLA BLA BLA
               calling some GL programme updating interface
        }
        
}
```
The "store" is registered into a root React dom instance through "mobx-react" "Provider" tag.

## expected result
The above function should be called every time the variable changed through HCI, and I can use them to update three.js engine codes

## actual result
The above *computed* functions are never called even with autorun. 
```js
react-dom.development.js:15374 Download the React DevTools for a better development experience: https://fb.me/react-devtools
app.js:49 test updateChoice
app.js:50 false
app.js:49 test updateChoice
app.js:50 true
app.js:49 test updateChoice
app.js:50 false
app.js:49 test updateChoice
app.js:50 true
```

# Why these happened and how should I organize my codes for better use os mobx to achieve my targets?

**Comment by mweststrate** [comment]
I spotted some obvious mistakes in your examples, can you fix those first? Beyond that I didn't really understand what you are actually asking :)

* `status.changed = !status.changed;`. Don't modify observables in computed properties. It creates an unendless loop (do you see why?). It is arbitrarily to decide what the value of `status.changed` should be after the computed is done, `true` or `false`.
* I get the impression that what you are looking for is simply: `reaction(()=> status.changed, changed => doSomethingUseful)` 
* `autorun(() => console.log(this.updateChoice));` That won't do anything ever, as it uses no observables, it will never run a second time. Do you mean something like `reaction(() => this.ModeChoice, (choice) => this.updateChoice())`?
* "This above function is called only once." Which function?
* I didn't understand this sentence: "I want to tell when a variable in tracked function is changed (not called)". What is the difference between changed and called and how does that relate to tracked functions? Which function?

**Comment by urugator** [comment]
**Problem 1**

Generally if you want to determine (compute) whether something changed you need to compare current and initial value. So you can have `@observable initialTodos` and `@computed onChange`, which performs the comparison of `todos` with `initialTodos`. 
To re-set the `onChange` status, move (current) `todos` to `initialTodos`.
Without initial value, changed status in uncomputable from current state and must be maintained as seperate state un-related to the rest. In that case it should be updated in actions (`addTodo`).
Updating changed status with reaction is possible, but at the cost of more `render` calls (or other dependent reactions).

Don't use `computed` to update state. It violates one way data flow (hard to follow, prone to errors), it may throw in strict mode, it may cause unwanted subscriptions (`completedTodosCount` subscribed to `status.changed`), it may cause staleness when `computed` operates in "non-reactive mode" (when not observed)...

**Problem 2**

Small interactive reproduction could help.

**Comment by yiakwy** [comment]
@urugator @mweststrate  I have updated my questions and fixed typos and thanks for your help !

I finished reading "what does mobx react to" and  "@computed" docs. I knew that mobx traces "arrows" (mutable value attributes). I just want to chain some behavior together, i.e.:

todos.add --modify mutable "todos"'s arrows--> computed value from "todos" used somewhere else -> computed function -> I know "todos" changed, I wan to modify other observables and automatically trigger functions associated to it. Hi, @urugator thanks for your reply! I just realize that your answer help me to understand another possible problem:

> if a observable's attribute changed from 0 to 1, how can I tell the difference in an "reactive" function ?  

Why I want it that way? In  the second example, I give an example that I want to do something in an reactive function: 

> Once I receive the signal an observable' attributes change, I will use "changed" signal to update other values

@mweststrate I think the problem here is that I didn't understand difference between "@computed" and "@autorun". Since my "computed" not tracked in react dom, it does not called.
 
Hi, @mweststrate  here is my updates:
1) This is my problem. Because the browser didn't throw an exception, and printed the results. It confused me for a while. I was wondering why it was just printed only once. Yes It brought endless loop if I could trigger another reactive function by modifying another observable attributes.

But what if I want to do something like
> When a observable attributes change, change another observable reactively

@urugator, @mweststrate , any suggestion from you is really appreciated. I just want to learn how to correctly, effectively use mobx. This problem coming into my mind when I played with MobX tutorial. Thanks for your tutorial, really good!

Is it logically correct?

3) typos I fixed it. You can see my output of "actual result" in Problem 2. I am sorry.
4) "This above function "isChanged" is called only once.". The browser didn't throw an exception for an endless loop but printed output from "isChanged" only once

**Comment by urugator** [comment]
> if a observable's attribute changed from 0 to 1, how can I tell the difference in an "reactive" function ?

Only the current state is available in reactive functions, not diffs.

> I know "todos" changed, I wan to modify other observables and automatically trigger functions associated to it.
> Once I receive the signal an observable' attributes change, I will use "changed" signal to update other values
> When a observable attributes change, change another observable reactively

Don't try to synchronize multiple observables. If something has to change, because something else changed, it should be represented as derivation (`@computed`).

```javascript
// when a changes, change b accordingly
@observable a;
@observable b; 

// => 

@observable a;
@computed get b() {
  return this.a + 5; 
}

// If b doesn't always change the same way, another state should represet the condition:
@observable a;
@observable c;
@computed get b() {
  if (this.c) {
    return this.a + 10;      
  }
  return this.a + 5; 
}
```

The following would work (not exactly the same way), but don't use it:
```javascript
// when a changes, change b accordingly
@observable a;
@observable b; 

reaction(() => this.a, () => this.b = this.a + 5);
```

Summary:
- modify observables in `actions`
- if some observable depends on another it should be `@computed`
- use `reaction` and `autorun` only to invoke async tasks (pesonally I suggest to forget that they exist)
- if some of it doesn't fit into your design, change your design

Example:
You have 2 windows and when you open one, you also want to open another:
```javascript
// Even though the following one-liner seems very convenient
reaction(() => window1.opened, () => window2.opened = window1.opened); 
// we won't use it, simply because the opening is synchronous operation (it's not the actual reason, but an easy way to recognize it shouldn't be used).

// For other reasons that are closely related to react philosophy and design, we won't use events:
class Window {
  @action open() {
    this.opened = true;
    this.emit("opened");
  }
}

window1.on("opened", () => window2.open());

// I believe there are 2 "proper" solutions (proper in way that they go hand in hand with react philosophy)

// 1) Second window is a computation of the first one
class Window2 {
  @computed opened() {
    return this.window1.opened;
  }
}
// Another version of the same is that you abstract the shared state and make both windows compution of this abstraction:
class Window {
  @computed opened() {
    return this.abstraction.opened;
  }
}   
// With this solution the state is propagated from bottom to top (unidirectionally)

// 2) Abstraction 
class Windows {
  @action open() {
    this.window1.open();
    this.window2.open();
  }
}
    
// With this solution the state is propagated from top to bottom (unidirectionally)
```

**Comment by mweststrate** [comment]
Closing as question seems answered. if not, feel free to comment and reopen
